### PR TITLE
[ADP-3306] Make more HTTP API errors machine-readable.

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] Convert to machine-readable error: `AssetNotPresent`
- [x] Convert to machine-readable error: `OutputTokenBundleSizeExceedsLimit`  
- [x] Convert to machine-readable error: `OutputTokenQuantityExceedsLimit`
- [x] Convert to machine-readable error: `TransactionAlreadyInLedger`
- [x] Convert to machine-readable error: `WrongEncryptionPassphrase` 

### Issue Number
ADP-3306

### CODE REVIEW

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 1]
```diff
@@ -356,23 +367,32 @@ instance IsServerError ErrMkTransaction where
...
+                    , pretty (view #address e)
+                    , ". Token policy identifier: "
+                    , pretty (view (#asset . #policyId) e)
+                    , ". Asset name: "
+                    , pretty (view (#asset . #assetName) e)
+                    , ". Token quantity specified: "
+                    , pretty (view #quantity e)
+                    , ". Maximum allowable token quantity: "
+                    , pretty (view #quantityMaxBound e)
+                    , "."
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]: Suggestion: can we delete this section?

Justification: all the relevant values are already included in the JSON record, so there's no need to repeat them in the error string.

```suggestion
```

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 2]
```diff
@@ -1021,38 +1045,53 @@ instance IsServerError ErrBalanceTxOutputError where
...
                 assetCount = TokenMap.size $
                     toWalletTokenBundle (snd output) ^. #tokens
+                errorMessage = T.unwords
+                    [ "One of the outputs you've specified contains too many "
+                    , "assets. Try splitting these assets across two or more "
+                    , "outputs. Destination address: "
+                    , pretty address
+                    , ". Asset count: "
+                    , pretty assetCount
+                    , "."
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]: Similar to https://github.com/cardano-foundation/cardano-wallet/pull/4622#discussion_r1634142800

```suggestion
                    , "outputs."
```

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 3]
```diff
@@ -1021,38 +1045,53 @@ instance IsServerError ErrBalanceTxOutputError where
...
+                    , pretty address'
+                    , ". Token policy identifier: "
+                    , pretty (show policyId)
+                    , ". Asset name: "
+                    , pretty (show assetName)
+                    , ". Token quantity specified: "
+                    , pretty (show quantity)
+                    , ". Maximum allowable token quantity: "
+                    , pretty (show quantityMaxBound)
+                    , "."
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]: Similar to https://github.com/cardano-foundation/cardano-wallet/pull/4622#discussion_r1634142800

```suggestion
```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 4]
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
         via DefaultRecord ApiErrorTxOutputLovelaceInsufficient
     deriving anyclass NFData
 
+data ApiErrorOutputTokenQuantityExceedsLimit = ApiErrorOutputTokenQuantityExceedsLimit
+    { address
+        :: !Text
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: Can we replace this with `ApiAnyAddress`?

```suggestion
        :: !ApiAnyAddress
```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 5]
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
...
+    , policyId
+        :: !Text
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: Request: use `TokenPolicyId` here.

```suggestion
        :: !(ApiT TokenPolicyId)
```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 6]
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
...
+    , assetName
+        :: !Text
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: Request: use `AssetName` here.
```suggestion
        :: !(ApiT AssetName)
```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 7]
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
...
+    deriving (Data, Eq, Generic, Show, Typeable)
+    deriving (FromJSON, ToJSON)
+        via DefaultRecord ApiErrorOutputTokenQuantityExceedsLimit
+    deriving anyclass NFData
+
+data ApiErrorOutputTokenBundleSizeExceedsLimit = ApiErrorOutputTokenBundleSizeExceedsLimit
+    { address
+        :: !Text
+    , bundleSize
+        :: !Int
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: Request: replace this with `Natural`

Justification: it's impossible to have a bundle that is negatively-sized.

```suggestion
        :: !Natural
```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 8]
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
...
 
+data ApiErrorOutputTokenQuantityExceedsLimit = ApiErrorOutputTokenQuantityExceedsLimit
+    { address
+        :: !Text
+    , policyId
+        :: !Text
+    , assetName
+        :: !Text
+    , quantity
+        :: !Natural
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: Request: use `TokenQuantity` here

```suggestion
        :: !(ApiT TokenQuantity)
```

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs [hunk 9]
```diff
@@ -293,6 +301,34 @@ data ApiErrorTxOutputLovelaceInsufficient = ApiErrorTxOutputLovelaceInsufficient
...
+    , maxQuantity
+        :: !Natural
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: Request: use `TokenQuantity` here

```suggestion
        :: !(ApiT TokenQuantity)
```

File: lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/CoinSelections.hs [hunk 10]
```diff
@@ -426,18 +426,17 @@ spec = describe "SHELLEY_COIN_SELECTION" $ do
...
+            sel <- makeRequest
+            verify sel [ expectResponseCode HTTP.status403 ]
+            decodeErrorInfo sel `Lifted.shouldBe`
+                OutputTokenQuantityExceedsLimit
+                ( ApiErrorOutputTokenQuantityExceedsLimit
+                    (toText $ apiAddress targetAddress)
+                    (toText policyId)
+                    (toText assetName)
+                    (unTokenQuantity excessiveQuantity)
+                    (unTokenQuantity txOutMaxTokenQuantity)
```
Comment (jonathanknowles) [inline lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/CoinSelections.hs]:
```suggestion
                    (apiAddress targetAddress)
                    policyId
                    assetName
                    excessiveQuantity
                    txOutMaxTokenQuantity
```

File: lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/CoinSelections.hs [hunk 11]
```diff
@@ -475,12 +474,11 @@ spec = describe "SHELLEY_COIN_SELECTION" $ do
...
-                        $ errMsg403OutputTokenBundleSizeExceedsLimit
-                            (apiAddress targetAddress)
-                            (assetCount)
-                    ]
+            sel <- makeRequest
+            verify sel [ expectResponseCode HTTP.status403 ]
+            decodeErrorInfo sel `Lifted.shouldBe`
+                OutputTokenBundleSizeExceedsLimit
+                ( ApiErrorOutputTokenBundleSizeExceedsLimit
+                    (toText $ apiAddress targetAddress)
```
Comment (jonathanknowles) [inline lib/integration/scenarios/Test/Integration/Scenario/API/Shelley/CoinSelections.hs:482]:
```suggestion
                    (apiAddress targetAddress)
```

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 12]
```diff
@@ -356,23 +367,32 @@ instance IsServerError ErrMkTransaction where
...
-                , pretty (view #quantityMaxBound e)
-                , "."
-                ]
+            flip (apiError err403) errorMessage $ OutputTokenQuantityExceedsLimit
+            ApiErrorOutputTokenQuantityExceedsLimit
+                { address = toText $ view #address e
+                , policyId = toText $ view (#asset . #policyId) e
+                , assetName = toText $ view (#asset . #assetName) e
+                , quantity = unTokenQuantity $ view #quantity e
+                , maxQuantity = unTokenQuantity $ view #quantityMaxBound e
```
Comment (jonathanknowles) [inline lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]:
```suggestion
                { address = view #address e
                , policyId = view (#asset . #policyId) e
                , assetName = view (#asset . #assetName) e
                , quantity = view #quantity e
                , maxQuantity = view #quantityMaxBound e
```

**Review by jonathanknowles — COMMENTED** [review]
Hi @paweljakubas 

Thanks for making this PR!

I have two general comments:

- **Suggestion:**
  Can we convert the error strings into _constants_ (i.e., stop building interpolated error strings)? Now that we have JSON objects to carry the variable parts of the error, I think there's no need to repeat this information within the error strings.
  
  _Example: https://github.com/cardano-foundation/cardano-wallet/pull/4622#discussion_r1634142800_
  
- **Request:**
  Within API error record types, can we replace the use of low-level types like `Text` and `Natural` with specialised types such as `ApiT TokenPolicyId`, `ApiT TokenQuantity`, `ApiAnyAddress`? The existing API error types all use specialised types.

  _Example: https://github.com/cardano-foundation/cardano-wallet/pull/4622#discussion_r1634144889_

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]: good point

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]: :+1:

File: lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Http/Server/Error.hs]: :+1:

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 5 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: done

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 6 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: done

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 7 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: done

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 8 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: done

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 9 above)
Comment (paweljakubas) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: done

File: lib/api/src/Cardano/Wallet/Api/Types/Error.hs (hunk 5 above)
Comment (paolino) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: This seems in conflict with the idea that `ApiT` is only used "outside" . But ok, the wallet is going in one direction the deposit wallet in another :whatever:

Comment (jonathanknowles) [reply lib/api/src/Cardano/Wallet/Api/Types/Error.hs]: > This seems in conflict with the idea that `ApiT` is only used "outside" . But ok, the wallet is going in one direction the deposit wallet in another :whatever:

Hi @paolino! I'm curious, what solution would you suggest here?

**Review by jonathanknowles — APPROVED** [review]
LGTM! 👍🏻 

I think it makes sense to defer making a decision on using a more dedicated type for the `address` fields for now. Switching to `ApiAddress n`, where `n` is a `NetworkDiscriminant` would require quite a few invasive changes.

**Pull request merged**